adapters/socketio_mbed_os5.c

```diff
@@ -0,0 +1,495 @@
+
+                socket_io_instance->on_io_error = on_io_error;
+                socket_io_instance->on_io_error_context = on_io_error_context;
+
+                socket_io_instance->io_state = IO_STATE_OPEN;
+
+                result = 0;
+            }
+        }
+    }
+    
+    if (on_io_open_complete != NULL)
+    {
+        on_io_open_complete(on_io_open_complete_context, result == 0 ? IO_OPEN_OK : IO_OPEN_ERROR);
+    }
+
+    return result;
+}
+
+int socketio_close(CONCRETE_IO_HANDLE socket_io, ON_IO_CLOSE_COMPLETE on_io_close_complete, void* callback_context)
+{
+    int result = 0;
+
+    if (socket_io == NULL)
+    {
+        result = __FAILURE__;
+    }
+    else
+    {
+        SOCKET_IO_INSTANCE* socket_io_instance = (SOCKET_IO_INSTANCE*)socket_io;
+        if (socket_io_instance->io_state == IO_STATE_CLOSED || 
+            socket_io_instance->io_state == IO_STATE_ERROR)
+        {
+            result = __FAILURE__;
+        }
+        else
+        {
+            close_tcp_connection(socket_io_instance);
+            if (on_io_close_complete != NULL)
+            {
+                on_io_close_complete(callback_context);
+            }
+            result = 0;
+        }
+    }
+    
+    return result;
+}
+
+int socketio_send(CONCRETE_IO_HANDLE socket_io, const void* buffer, size_t size, ON_SEND_COMPLETE on_send_complete, void* callback_context)
+{
+    int result;
+
+    if ((socket_io == NULL) ||
+        (buffer == NULL) ||
+        (size == 0))
+    {
+        /* Invalid arguments */
+        result = __FAILURE__;
+    }
+    else
+    {
+        result = 0;
+
+        SOCKET_IO_INSTANCE* socket_io_instance = (SOCKET_IO_INSTANCE*)socket_io;
+        if (socket_io_instance->io_state != IO_STATE_OPEN)
+        {
+            result = __FAILURE__;
+        }
+        else
+        {
+            // Queue the data, and the socketio_dowork sends the package later
+            if (add_pending_io(socket_io_instance, buffer, size, on_send_complete, callback_context) != 0)
+            {
+                result = __FAILURE__;
+            }
+            
+            if (send_queued_data(socket_io_instance) < 0)
+            {
+                result = __FAILURE__;
+            }
+        }
+    }
+
+    return result;
+}
+
+void socketio_dowork(CONCRETE_IO_HANDLE socket_io)
+{
+    if (socket_io != NULL)
+    {
+        SOCKET_IO_INSTANCE* socket_io_instance = (SOCKET_IO_INSTANCE*)socket_io;
+        if (socket_io_instance->io_state == IO_STATE_OPEN)
+        {
+            // Retrieve all data from IoT Hub
+            if (retrieve_data(socket_io_instance) < 0)
+            {
+                return;
+            }
+        
+            // Send all packages in the queue
+            send_queued_data(socket_io_instance);
+        }
+    }
+}
+
+int socketio_setoption(CONCRETE_IO_HANDLE socket_io, const char* optionName, const void* value)
+{
+    /* Not implementing any options, do nothing */
+    return OPTIONHANDLER_OK;
```

Comment:
>OPTIONHANDLER_OK [](start = 11, length = 16)

Please don't silently fail. Return __FAILURE__

Reply:
return FAILURE will cause failed at the beginning during init the client, here there is no actions for mbedos, so just leave it as empty.